cmake/AwsGetVersion.cmake

```diff
@@ -0,0 +1,58 @@
+function(aws_get_version var_version_simple var_version_full var_git_hash)
+    # Simple version is "MAJOR.MINOR.PATCH" from VERSION file
+    file(READ "${CMAKE_CURRENT_SOURCE_DIR}/VERSION" version_simple)
+    string(STRIP ${version_simple} version_simple)
+    set(${var_version_simple} ${version_simple} PARENT_SCOPE)
+
+    # By default, full version is same as simple version.
+    # But we'll make it more specific later, if we determine that we're not at an exact tagged commit.
+    set(${var_version_full} ${version_simple} PARENT_SCOPE)
+
+    # Get git hash
+    aws_git_try("rev-parse HEAD" git_hash git_success)
+    if (git_success)
+        set(${var_git_hash} ${git_hash} PARENT_SCOPE)
+
+        # Determine if we're at the exact tagged commit
+        set(is_exact_version FALSE)
+        aws_git_try("tag --points-at HEAD" head_tags git_success)
+        if (git_success)
+            foreach(tag IN LISTS head_tags) # cmake 3.3+ could use IN_LIST instead
+                if (tag STREQUAL "v${version_simple}")
+                    set(is_exact_version TRUE)
+                endif()
+            endforeach()
+        endif()
+
+        # Full version should indicate when we're not at the exact tagged commit.
+        # Be compliant with https://semver.org
+        if (NOT is_exact_version)
+            aws_git_try("rev-parse --short=8 HEAD" git_hash_short git_success)
+            if (git_success)
+                set(${var_version_full} "${version_simple}-dev+${git_hash_short}" PARENT_SCOPE)
+            endif()
+        endif()
```

Comment:
An eternal question...

Yeah, if variable is mostly used for logging, then a fallback like "UnknownGitHash" is nice. Someone can blindly log it, and the logs are always readable

Given: `printf("CRT %d.%d.%d - %s", MAJOR, MINOR, PATCH, HASH);`
"CRT 1.2.3 - UnknownGitHash"
is nicer than
"CRT 1.2.3 - "

But if someone is writing logic that branches on whether or not this is set, then something like empty string is nicer than an arbitrary string value. A customer would rather write:
`if (strlen(AWS_CRT_CPP_GIT_HASH) > 0)`
than
`if (strcmp(AWS_CRT_CPP_GIT_HASH, "UnknownGitHash") != 0)`

Reply:
I'm going to keep AWS_CRT_CPP_GIT_HASH blank, since AWS_CRT_CPP_VERSION is already the "nice for logging" string, and includes the short git hash when it's important